AX: content-visibility: hidden does not remove elements from the accessibility tree

twilco opened the pull request:
#### 3727ac7f682278095aecd5714957126d38d72b13
<pre>
AX: content-visibility: hidden does not remove elements from the accessibility tree
<a href="https://bugs.webkit.org/show_bug.cgi?id=285955">https://bugs.webkit.org/show_bug.cgi?id=285955</a>
<a href="https://rdar.apple.com/142915090">rdar://142915090</a>

Reviewed by Chris Fleizach.

When an element has content-visibility:hidden, it should not be exposed in the accessibility tree. This commit fixes that.

This bug caused us to expose collapsed content in details elements, which is hidden with content-visibility:hidden after
<a href="https://github.com/WebKit/WebKit/commit/cd2f382c7dc0e50ae049f7e2420377e944ba60a8.">https://github.com/WebKit/WebKit/commit/cd2f382c7dc0e50ae049f7e2420377e944ba60a8.</a>

This patch was co-authored with Josh Hoffman (<a href="https://github.com/hoffmanjoshua).">https://github.com/hoffmanjoshua).</a>

* LayoutTests/accessibility/details-summary-content-hidden-expected.txt: Added.
* LayoutTests/accessibility/details-summary-content-hidden.html: Added.
* LayoutTests/accessibility/mac/details-summary-expected.txt:
* LayoutTests/accessibility/mac/details-summary.html:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::isNodeVisible const):
(WebCore::AXObjectCache::onStyleChange):
(WebCore::AXObjectCache::handleAttributeChange):
(WebCore::isVisibilityHidden):
(WebCore::isRenderHidden):
(WebCore::AXObjectCache::onExpandedChanged): Deleted.
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::defaultObjectInclusion const):
(WebCore::AccessibilityObject::isWithinHiddenWebArea const):
* Source/WebCore/html/HTMLDetailsElement.cpp:
(WebCore::HTMLDetailsElement::toggleOpen): Deleted.
* Source/WebCore/html/HTMLDetailsElement.h:

Canonical link: <a href="https://commits.webkit.org/288940@main">https://commits.webkit.org/288940@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8ac1069356276f9a48b3e131084420514d999456

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84745 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/4470 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39133 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/89884 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/35797 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/86830 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/4559 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/12445 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/65946 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23776 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87790 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/3458 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77008 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46220 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/3337 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/31227 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/34871 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/74188 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/32033 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/91260 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/12084 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/8823 "Found 2 new test failures: http/wpt/webrtc/third-party-frame-ice-candidate-filtering.html swipe/navigate-event-back-swipe-verify-ua-transition.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/74423 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/12311 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/72819 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73549 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18216 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/17923 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/16366 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/3532 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/12036 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/17476 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/11870 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/15364 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/13616 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hoffmanjoshua commented on LayoutTests/accessibility/details-summary-content-hidden-expected.txt:
```diff
@@ -0,0 +1,16 @@
+This tests that the summary content inaccessible when a details element is collapsed.
```
Missing "is" before inaccessible here.

hoffmanjoshua commented on LayoutTests/accessibility/details-summary-content-hidden.html:
```diff
@@ -0,0 +1,45 @@
+
+    output += expect("details.isExpanded", "false");
+    output += expect("summary.isExpanded", "false");
+    output += expect("!accessibilityController.accessibleElementById('content')", "true")
```
I think I wrote this, but I missed a ";" here (and also on lines 31 and 36)

hoffmanjoshua reviewed (commented):
r+

twilco committed:
AX: content-visibility: hidden does not remove elements from the accessibility tree
https://bugs.webkit.org/show_bug.cgi?id=285955
rdar://142915090

Reviewed by Chris Fleizach.

When an element has content-visibility:hidden, it should not be exposed in the accessibility tree. This commit fixes that.

This bug caused us to expose collapsed content in details elements, which is hidden with content-visibility:hidden after
https://github.com/WebKit/WebKit/commit/cd2f382c7dc0e50ae049f7e2420377e944ba60a8.

This patch was co-authored with Josh Hoffman (https://github.com/hoffmanjoshua).

* LayoutTests/accessibility/details-summary-content-hidden-expected.txt: Added.
* LayoutTests/accessibility/details-summary-content-hidden.html: Added.
* LayoutTests/accessibility/mac/details-summary-expected.txt:
* LayoutTests/accessibility/mac/details-summary.html:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::isNodeVisible const):
(WebCore::AXObjectCache::onStyleChange):
(WebCore::AXObjectCache::handleAttributeChange):
(WebCore::isVisibilityHidden):
(WebCore::isRenderHidden):
(WebCore::AXObjectCache::onExpandedChanged): Deleted.
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::defaultObjectInclusion const):
(WebCore::AccessibilityObject::isWithinHiddenWebArea const):
* Source/WebCore/html/HTMLDetailsElement.cpp:
(WebCore::HTMLDetailsElement::toggleOpen): Deleted.
* Source/WebCore/html/HTMLDetailsElement.h:

Canonical link: https://commits.webkit.org/288940@main
